src/details/ArborX_MinimumSpanningTree.hpp

```diff
@@ -0,0 +1,504 @@
+    // edge
+    return KokkosExt::min(component, next_component);
+  }
+
+  KOKKOS_FUNCTION auto computeFinalComponent(int component) const
+  {
+    int prev_component = component;
+    int next_component;
+    while ((next_component = computeNextComponent(prev_component)) !=
+           prev_component)
+      prev_component = next_component;
+
+    return next_component;
+  }
+
+  KOKKOS_FUNCTION void operator()(int i) const
+  {
+    auto const component = labels_(i);
+    auto const final_component = computeFinalComponent(component);
+    labels_(i) = final_component;
+    if (i != component)
+    {
+      return;
+    }
+    auto const n = out_edges_.extent(0);
+    if (i != final_component)
+    {
+      auto const edge = out_edges_(i - n + 1);
+      // append new edge at the "end" of the array (akin to
+      // std::vector::push_back)
+      auto const back =
+          Kokkos::atomic_fetch_inc(&num_edges_()); // atomic post increment
+      edges_(back) = edge;
+    }
+  }
+};
+
+// Merge components and append new edges
+template <class ExecutionSpace, class Labels, class ComponentOutEdges,
+          class Edges, class EdgesCount>
+void updateComponentsAndEdges(ExecutionSpace const &space,
+                              ComponentOutEdges const &component_out_edges,
+                              Labels const &labels, Edges const &edges,
+                              EdgesCount const &num_edges)
+{
+  UpdateComponentsAndEdges<Labels, ComponentOutEdges, Edges, EdgesCount>(
+      space, labels, component_out_edges, edges, num_edges);
+}
+
+// Reverse node leaf permutation order back to original indices
+template <class ExecutionSpace, class BVH, class Edges>
+void finalizeEdges(ExecutionSpace const &space, BVH const &bvh,
+                   Edges const &edges)
+{
+  auto const n = bvh.size();
+  ARBORX_ASSERT(edges.extent(0) == n - 1);
+  Kokkos::parallel_for(
+      "ArborX::MST::finalize_edges",
+      Kokkos::RangePolicy<ExecutionSpace>(space, 0, n - 1),
+      KOKKOS_LAMBDA(int i) {
+        edges(i).source =
+            HappyTreeFriends::getLeafPermutationIndex(bvh, edges(i).source);
+        edges(i).target =
+            HappyTreeFriends::getLeafPermutationIndex(bvh, edges(i).target);
+      });
+}
+
+template <class ExecutionSpace, class BVH, class Labels, class Metric,
+          class Radii>
+void resetSharedRadii(ExecutionSpace const &space, BVH const &bvh,
+                      Labels const &labels,
+                      Metric const &mutual_reachability_distance,
+                      Radii const &radii)
+{
+  auto const n = bvh.size();
+  Kokkos::parallel_for(
+      "ArborX::MST::reset_shared_radii",
+      Kokkos::RangePolicy<ExecutionSpace>(space, n - 1, 2 * n - 2),
+      KOKKOS_LAMBDA(int i) {
+        int const j = i + 1;
+        auto const label_i = labels(i);
+        auto const label_j = labels(j);
+        if (label_i != label_j)
+        {
```

Comment:
Needs a big fat comment here to explain. Something along the lines:
```
We will search for the shortest outgoing edge of a component. The better we initialize
the upper bound on the distance (i.e., the smaller it is), the less traversal we will do and
the faster it will be.

Here, we use the knowledge that it is a self-collision problem. In other words, we only
have a single point cloud. We further use the fact that if we sort predicates based on
the Morton codes, it will match the order of predicates (or be close enough, as some
points with the same Morton codes may be in a different order due to the unstable sort
that we use). Thus, if we take an index of a query, we assume that it matches corresponding
primitive. If a label of that primitive is different from a label of its neighbor (which is
in fact its Morton neighbor), we compute the distance between the two. The upper bound for
a component is set to the minimum distance between all such pairs. Given that the
Morton neighbors are typically close to each other, this should provide a reasonably low bound.
```